README.md
[![Build Status](https://secure.travis-ci.org/STRML/node-toobusy.png)](http://travis-ci.org/STRML/node-toobusy)

# Is Your Node Process Too Busy?

`toobusy-js` is a fork of lloyd's [node-toobusy](http://github.com/lloyd/node-toobusy) that removes native dependencies
in favor of using the `unref` introduced in [node 0.9.1](http://blog.nodejs.org/2012/08/28/node-v0-9-1-unstable/).

This package is a simpler install without native dependencies, but requires node >= 4.0.0.

## Node-Toobusy

What happens when your service is overwhelmed with traffic?
Your server can do one of two things:

  * Stop working, or...
  * Keep serving as many requests as possible

This library helps you do the latter.

## How it works

`toobusy` polls the node.js event loop and keeps track of "lag",
which is long requests wait in node's event queue to be processed.
When lag crosses a threshold, `toobusy` tells you that you're *too busy*.
At this point you can stop request processing early
(before you spend too much time on them and compound the problem),
and return a "Server Too Busy" response.
This allows your server to stay *responsive* under extreme load,
and continue serving as many requests as possible.

## installation

``` sh-session
npm install toobusy-js
```


## usage

```js
let toobusy = require('toobusy-js'),
    express = require('express');

let app = express();

// middleware which blocks requests when we're too busy
app.use(function(req, res, next) {
  if (toobusy()) {
    res.status(503).send("I'm busy right now, sorry.");
  } else {
    next();
  }
});

app.get('/', function(req, res) {
  // processing the request requires some work!
  let i = 0;
  while (i < 1e5) i++;
  res.send("I counted to " + i);
});

let server = app.listen(3000);

process.on('SIGINT', function() {
  server.close();
  // calling .shutdown allows your process to exit normally
  toobusy.shutdown();
  process.exit(0);
});
```

## tunable parameters

The one knob that the library exposes is "maximum lag".
This number represents the maximum amount of time in milliseconds that the event queue is behind,
before we consider the process *too busy*.

```js
let toobusy = require('toobusy-js');

// Set maximum lag to an aggressive value.
toobusy.maxLag(10);

// Set check interval to a faster value. This will catch more latency spikes
// but may cause the check to be too sensitive.
toobusy.interval(250);

// Get current maxLag or interval setting by calling without parameters.
let currentMaxLag = toobusy.maxLag(), interval = toobusy.interval();
```

The default maxLag value is 70ms, and the default check interval is 500ms.
This allows an "average" server to run at 90-100% CPU
and keeps request latency at around 200ms.
For comparison, a maxLag value of 10ms results in 60-70% CPU usage,
while latency for "average" requests stays at about 40ms.

These numbers are only examples,
and the specifics of your hardware and application can change them drastically,
so experiment!
The default of 70 should get you started.

## references

> There is nothing new under the sun. (Ecclesiastes 1:9)

Though applying "event loop latency" to node.js was not directly inspired by anyone else's work,
this concept is not new.  Here are references to others who apply the same technique:

  * [Provos, Lever, and Tweedie 2000](http://www.kegel.com/c10k.html#tips) - "notes that dropping incoming connections when the server is overloaded improved the shape of the performance curve."

## license

[WTFPL](http://wtfpl.org)


ChangeLog
0.4.2
  * version bump

0.4.1
  * version bump

0.4.0
  * refactoring
  * new tests
  * added `toobusy.interval()`
  * many comments

0.3.1
  * fix package.json plural for repository

0.3.0 -
  * full JS implementation, removed c code & bindings
  * example bugfixes

0.2.3 -
  * node.js 0.10.0 support - uprade to latest bindings and include in .travis.yml

0.2.2 -
  * *really* works on windows
  * improved documentation

0.2.1 -
  * works on windows
  * improved express.js example (now targets express 3.0)

0.2.0 -
  * documentation improvements
  * improved examples (including load generation tool and server)
  * .lag() added to retrieve (periodically calculated) event loop lag in ms

0.1.1 -
  * hey look, unit tests!

0.1.0 -
  * improved documentation and samples
  * added `maxLag` parameter


tests.js
'use strict';
var should = require('should');
var toobusy = require('./');

function tightWork(duration) {
  var start = Date.now();
  while ((Date.now() - start) < duration) {
    for (var i = 0; i < 1e5;) i++;
  }
}

/*global describe, it, beforeEach, afterEach */
describe('the library', function() {
  it('should export a couple functions', function() {
    (toobusy).should.be.a.function;
    (toobusy.maxLag).should.be.a.function;
    (toobusy.shutdown).should.be.a.function;
    (toobusy.interval).should.be.a.function;
    (toobusy.shutdown).should.be.a.function;
    (toobusy).should.not.have.property('start');
  });
  it('should start automatically', function() {
    (toobusy.started()).should.equal(true);
  });
});

describe('maxLag', function() {
  it('should default to 70', function() {
    (toobusy.maxLag()).should.equal(70);
  });
  it('should throw an exception for non-numbers', function() {
    (function() { toobusy.maxLag('derp'); }).should.throw(/must be a number/);
  });
  it('should throw an exception for values < 10', function() {
    (function() { toobusy.maxLag(9); }).should.throw(/should be greater than 10/);
  });
  it('should be configurable', function() {
    (toobusy.maxLag(50)).should.equal(50);
    (toobusy.maxLag(10)).should.equal(10);
    (toobusy.maxLag()).should.equal(10);
  });
});

describe('interval', function() {
  it('should default to 500', function() {
    (toobusy.interval()).should.equal(500);
  });
  it('should throw an exception for values < 16', function() {
    (function() { toobusy.interval(15); }).should.throw(/Interval/);
  });
  it('should be configurable', function() {
    (toobusy.interval(250)).should.equal(250);
    (toobusy.interval(300)).should.equal(300);
    (toobusy.interval()).should.equal(300);
  });
});

describe('toobusy()', function() {
  // Set lower thresholds for each of these tests.
  // Resetting the interval() also resets the internal lag counter, which
  // is nice for making these tests independent of each other.
  beforeEach(function() {
    toobusy.maxLag(10);
    toobusy.interval(250);
  });
  afterEach(function() {
    toobusy.maxLag(70);
    toobusy.interval(500);
  });
  it('should return true after a little load', function(done) {
    function load() {
      if (toobusy()) return done();
      tightWork(100);
      setTimeout(load, 0);
    }
    load();
  });

  it('should return a lag value after a little load', function(done) {
    function load() {
      if (toobusy()) {
        var lag = toobusy.lag();
        should.exist(lag);
        lag.should.be.above(1);
        return done();
      }
      tightWork(100);
      setTimeout(load, 0);
    }
    load();
  });
});

describe('smoothingFactor', function() {
  beforeEach(function() {
    toobusy.maxLag(10);
    toobusy.interval(250);
  });
  afterEach(function() {
    toobusy.maxLag(70);
    toobusy.interval(500);
  });
  it('should default to 1/3', function(done) {
    (toobusy.smoothingFactor()).should.equal(1/3);
    done();
  });
  it('should throw an exception for invalid values', function(done) {
    (function() { toobusy.smoothingFactor(0); }).should.throw;
    (function() { toobusy.smoothingFactor(2); }).should.throw;
    (function() { toobusy.smoothingFactor(-1); }).should.throw;
    (function() { toobusy.smoothingFactor(1); }).should.not.throw;
    done();
  });
  it('should be configurable', function(done) {
    (toobusy.smoothingFactor(0.9)).should.equal(0.9);
    (toobusy.smoothingFactor(0.1)).should.equal(0.1);
    (toobusy.smoothingFactor()).should.equal(0.1);
    done();
  });
  it('should allow no dampening', function(done) {
    var cycles_to_toobusy = 0;
    toobusy.smoothingFactor(1); // no dampening

    function load() {
      if (toobusy()) {
        (cycles_to_toobusy).should.equal(3);
        return done();
      }
      cycles_to_toobusy++;
      tightWork(100); // in 3 ticks, will overshoot by ~50ms, above 2*10ms
      setImmediate(load);
    }

    load();
  });
  it('should respect larger dampening factors', function(done) {
    var cycles_to_toobusy = 0;
    toobusy.smoothingFactor(0.05);

    function load() {
      if (toobusy()) {
        (cycles_to_toobusy).should.be.above(3);
        return done();
      }
      cycles_to_toobusy++;
      tightWork(100);
      setImmediate(load);
    }

    load();
  });
});

describe('started', function() {
  it('should return false after shutdown', function(done) {
    toobusy.shutdown();
    (toobusy.started()).should.equal(false);
    done();
  });
});




examples/http.js
'use strict';

let http = require('http'),
 toobusy = require('..');

console.log("Maximum allowed event loop lag: " + toobusy.maxLag(50) + "ms");

function processRequest(res, num, startTime) {
  if (!startTime) startTime = new Date();

  if (num === undefined) {
    return process.nextTick(function() {
      processRequest(res, 0);
    });
  }
  if (num >= 5) {
    res.writeHead(200, {'Content-Type': 'text/plain'});
    res.end('I counted to ' + num + ' in ' + (new Date() - startTime) + 'ms\n');
  } else {
    // 1ms of computation
    let targetTime = (new Date() - startTime) + 1;
    while (new Date() - startTime < targetTime);
    processRequest(res, num + 1, startTime);
  }
}

http.createServer(function (req, res) {
  if (toobusy()) {
    res.writeHead(503);
    return res.end();
  }

  // we're not too busy!  let's process a request!
  processRequest(res);
}).listen(3000, '127.0.0.1', 2048);


examples/load.js
'use strict';

let http = require('http');

// a little load generate client that generates constant load over
// even if the server cannot keep up

let running = 0;
let twoHundred = 0;
let fiveOhThree = 0;
let yucky = 0;
let avg = 0;

// how many requests per second should we run?
const rps = (process.env['RPS'] || 40) / 40;
let curRPS = rps;
let started = 0;
const startTime = new Date();
let lastMark = startTime;

let ivalnum = 0;

setInterval(function() {
  ivalnum++;
  function startOne() {
    started++;
    running++;
    let start = new Date();
    let endOrError = false;
    function cEndOrError() {
      if (endOrError) console.log("end AND error");
      endOrError = true;
    }
    let req = http.get({
      host: '127.0.0.1',
      port: 3000,
      agent: false,
      path: '/',
      headers: {
        "connection": "close"
      }
    }, function(res) {
      if (res.statusCode === 503) {
        fiveOhThree++;
      } else {
        twoHundred++;
      }
      avg = ((new Date() - start) + avg * started) / (started + 1);
      running--;
      cEndOrError();
    }).on('error', function(e) {
      process.stderr.write(e.toString() + " - " + (new Date() - start) + "ms\n");
      avg = ((new Date() - start) + avg * started) / (started + 1);
      running--;
      yucky++;
      cEndOrError();
    });
  }

  for (let i = 0; i < curRPS ; i++) startOne();

  // report and scale up every 2s
  if (!(ivalnum % (40 * 2))) {
    let delta = (new Date() - lastMark) / 1000.0 ;
    console.log(Math.round((new Date() - startTime) / 1000.0),
                Math.round(started / delta),
                Math.round(twoHundred / delta),
                Math.round(fiveOhThree / delta),
                avg,
                Math.round(yucky / delta));
    curRPS = curRPS + .5;
    started = twoHundred = fiveOhThree = yucky = 0;
    lastMark = new Date();
  }
}, 25);


toobusy.js
'use strict';

//
// Constants
//
const STANDARD_HIGHWATER = 70;
const STANDARD_INTERVAL = 500;

// A dampening factor.  When determining average calls per second or
// current lag, we weigh the current value against the previous value 2:1
// to smooth spikes.
// See https://en.wikipedia.org/wiki/Exponential_smoothing
const SMOOTHING_FACTOR = 1/3;

//
// Vars
//

let lastTime = Date.now();
let highWater = STANDARD_HIGHWATER;
let interval = STANDARD_INTERVAL;
let smoothingFactor = SMOOTHING_FACTOR;
var currentLag = 0;
let checkInterval;


/**
 * Main export function.
 * @return {Boolean} True if node process is too busy.
 */
let toobusy = function(){
  // If current lag is < 2x the highwater mark, we don't always call it 'too busy'. E.g. with a 50ms lag
  // and a 40ms highWater (1.25x highWater), 25% of the time we will block. With 80ms lag and a 40ms highWater,
  // we will always block.
  let pctToBlock = (currentLag - highWater) / highWater;
  return Math.random() < pctToBlock;
};

/**
 * Sets or gets the current check interval.
 * If you want more sensitive checking, set a faster (lower) interval. A lower maxLag can also create a more
 * sensitive check.
 * @param  {Number} [newInterval] New interval to set. If not provided, will return the existing interval.
 * @return {Number}               New or existing interval.
 */
toobusy.interval = function(newInterval) {
  if (!newInterval) return interval;
  if (typeof newInterval !== "number") throw new Error("Interval must be a number.");

  newInterval = Math.round(newInterval);
  if(newInterval < 16) throw new Error("Interval should be greater than 16ms.");

  toobusy.shutdown();
  interval = newInterval;
  start();
  return interval;
};

/**
 * Returns last lag reading from last check interval.
 * @return {Number} Lag in ms.
 */
toobusy.lag = function(){
  return Math.round(currentLag);
};

/**
 * Set or get the current max latency threshold. Default is 70ms.
 *
 * Note that if event loop lag goes over this threshold, the process is not always 'too busy' - the farther
 * it goes over the threshold, the more likely the process will be considered too busy.
 *
 * The percentage is equal to the percent over the max lag threshold. So 1.25x over the maxLag will indicate
 * too busy 25% of the time. 2x over the maxLag threshold will indicate too busy 100% of the time.
 * @param  {Number} [newLag] New maxLag (highwater) threshold.
 * @return {Number}          New or existing maxLag (highwater) threshold.
 */
toobusy.maxLag = function(newLag){
  if(!newLag) return highWater;

  // If an arg was passed, try to set highWater.
  if (typeof newLag !== "number") throw new Error("MaxLag must be a number.");
  newLag = Math.round(newLag);
  if(newLag < 10) throw new Error("Maximum lag should be greater than 10ms.");

  highWater = newLag;
  return highWater;
};

/**
 * Set or get the smoothing factor. Default is 0.3333....
 *
 * The smoothing factor per the standard exponential smoothing formula "αtn + (1-α)tn-1"
 * See: https://en.wikipedia.org/wiki/Exponential_smoothing
 *
 * @param  {Number} [newFactor] New smoothing factor.
 * @return {Number}             New or existing smoothing factor.
 */
toobusy.smoothingFactor = function(newFactor){
  if(!newFactor) return smoothingFactor;

  if (typeof newFactor !== "number") throw new Error("NewFactor must be a number.");
  if(newFactor <= 0 || newFactor > 1) throw new Error("Smoothing factor should be in range ]0,1].");

  smoothingFactor = newFactor;
  return smoothingFactor;
};

/**
 * Shuts down toobusy.
 *
 * Not necessary to call this manually, only do this if you know what you're doing. `unref()` is called
 * on toobusy's check interval, so it will never keep the server open.
 */
toobusy.shutdown = function(){
  currentLag = 0;
  checkInterval = clearInterval(checkInterval);
};

toobusy.started = function() {
  return !!checkInterval;
};

/**
 * Private - starts checking lag.
 */
function start() {
  checkInterval = setInterval(function(){
    let now = Date.now();
    let lag = now - lastTime;
    lag = Math.max(0, lag - interval);
    // Dampen lag. See SMOOTHING_FACTOR initialization at the top of this file.
    currentLag = smoothingFactor * lag + (1 - smoothingFactor) * currentLag;
    lastTime = now;
  }, interval);

  // Don't keep process open just for this timer.
  checkInterval.unref();
}

// Kickoff the checking!
start();

module.exports = toobusy;
